rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,499 @@
+        }
+    }
+
+    unsafe fn thread_backward_pointer_fields(&mut self, object: *mut Obj) {
+        debug_assert!(object.tag() < TAG_ARRAY_SLICE_MIN);
+        visit_pointer_fields(
+            &mut (),
+            object,
+            object.tag(),
+            self.heap.limits.base,
+            |_, field_address| {
+                let field_value = *field_address;
+                // Thread if backwards or self pointer
+                if field_value.get_ptr() <= object as usize {
+                    (&self).thread(field_address);
+                }
+            },
+            |_, _, array| array.len(),
+        );
+    }
+
+    // Thread forward pointers in old generation leading to young generation
+    unsafe fn thread_old_generation_pointers(&mut self) {
+        let mut iterator = REMEMBERED_SET.as_ref().unwrap().iterate();
+        while iterator.has_next() {
+            let location = iterator.current().get_raw() as *mut Value;
+            let value = *location;
+            // value in the location may have changed since recording by the write barrer
+            if value.points_to_or_beyond(self.heap.limits.last_free) {
+                debug_assert!((location as usize) >= self.heap.limits.base);
+                self.thread(location);
+            }
+            iterator.next();
+        }
+    }
+
+    unsafe fn move_phase(&mut self) {
+        REMEMBERED_SET = None; // no longer valid when the moving phase starts
+        let mut free = self.heap.limits.base;
+
+        let mut bitmap_iter = iter_bits();
+        if self.strategy == Strategy::Young {
+            bitmap_iter.advance(self.heap.limits.last_free as u32);
+            free = self.heap.limits.last_free;
+        }
+        let mut bit = bitmap_iter.next();
+        while bit != BITMAP_ITER_END {
+            let old_pointer = (bit * WORD_SIZE) as *mut Obj;
+            let new_pointer = free;
+
+            // Update backwards references to the object's new location and restore object header
+            self.unthread(old_pointer, new_pointer);
+
+            // Move the object
+            let object_size = object_size(old_pointer as usize);
+            if new_pointer as usize != old_pointer as usize {
+                memcpy_words(new_pointer as usize, old_pointer as usize, object_size);
+                debug_assert!(object_size.as_usize() > size_of::<Obj>().as_usize());
+            }
+
+            free += object_size.to_bytes().as_usize();
+
+            // Thread forward pointers of the object, even if not moved
+            self.thread_forward_pointers(new_pointer as *mut Obj);
+
+            bit = bitmap_iter.next();
+        }
+
+        self.heap.limits.free = free;
+    }
+
+    /// Thread forward pointers in object
+    unsafe fn thread_forward_pointers(&mut self, object: *mut Obj) {
+        visit_pointer_fields(
+            &mut (),
+            object,
+            object.tag(),
+            self.heap.limits.base,
```

Comment:
Again, for a young collection, can we restrict this to `heap.limits.last_free`?
Similar to previous comment.

Reply:
Also improved in [fcbb1a4](https://github.com/dfinity/motoko/pull/3495/commits/fcbb1a434dac240daaac1e059804a3cfaa3f1459).